internal/configs/virtualserver.go

```diff
@@ -198,15 +198,24 @@ func generateVirtualServerConfig(virtualServerEx *VirtualServerEx, tlsPemFileNam
 	}
 }
 
-func generateUpstream(upstreamName string, upstream conf_v1alpha1.Upstream, endpoints []string, isPlus bool, cfgParams *ConfigParams) version2.Upstream {
+func generateUpstream(upstreamName string, upstream conf_v1alpha1.Upstream, virtualServerEx *VirtualServerEx, isPlus bool, isResolverConfigured bool, cfgParams *ConfigParams) version2.Upstream {
```

Comment:
I don't think ` virtualServerEx *VirtualServerEx` is neither required nor at the right level of abstractions for `generateUpstream`. I know we have something similar in the old code (like `createUpstream` in `ingress.go`). However, for the new code, I would rather not use the old code as a good example. 

As a bad consequece of this change, a lot of unnecessary test input data was created in generateUpstream tests. For example:
```
upstream := conf_v1alpha1.Upstream{Service: name, Port: 80}
	vsEx := &VirtualServerEx{
		VirtualServer: &conf_v1alpha1.VirtualServer{
			ObjectMeta: meta_v1.ObjectMeta{
				Name:      name,
				Namespace: "test-namespace",
			},
		},
		Endpoints: map[string][]string{
			"test-namespace/test-upstream:80": []string{
				"192.168.10.10:8080",
			},
		},
	}
```
Perhaps two arguments in this case - `endpoints[]string` and `externalNameSvcs map[string]bool` would be more suitable for `generateUpstream` instead of `virtualServerEx *VirtualServerEx`. While we have 2 arguments instead of one, both arguments are required by `generateUpstream` and at the right level of abstraction.

Reply:
Too many arguments, I don't see any problem with using the `virtualServerEx` Actually I jsut realized I need to pass the namespace too, cos is different for VS and VSR, so, either we have different functions for both (this could be a good use of Generics when they come to go), or I use `virtualServerEx`, otherwise, it will be too many arguments.

I understand tests are part of the code, but, if I have to choose between simplicity in code or tests I'll go with code all the time. You spend more time with actual code than tests. 

Unless you really thing this is super important, I'll go with `virtualServerEx`.

Reply:
Alternatively, we can create a Struct like `UpstreamConfig` or something, and just use the important fields there.

Reply:
Thanks for the explanation.

What if we use another approach and put the logic related to transforming the endpoints in a separate function like this (not tested, just a prototype):
```go
func prepareEndpoints(endpoints []string, isExternalNameSvc bool, isResolverConfigured bool, isPlus bool, service string) []string {
	if !isPlus && len(endpoints) == 0 {
		return []string{nginx502Server}
	}

	if isExternalNameSvc && !isResolverConfigured {
		glog.Warningf("A resolver must be configured for Type ExternalName service %s, no upstream servers will be created", service)
		return []string{}
	}

	return endpoints
}
```

In `generateVirtualServerConfig` will have:
```go
		upstreamName := virtualServerUpstreamNamer.GetNameForUpstream(u.Name)
		endpointsKey := GenerateEndpointsKey(virtualServerEx.VirtualServer.Namespace, u.Service, u.Port)
		endpoints := virtualServerEx.Endpoints[endpointsKey]
		endpoints = prepareEndpoints(endpoints, virtualServerEx.ExternalNameSvcs[endpointsKey], isResolverConfigured, isPlus, u.Service)
		upstreamNamespace := virtualServerEx.VirtualServer.Namespace
		ups := generateUpstream(upstreamName, u, endpoints, virtualServerEx.ExternalNameSvcs[endpointsKey], baseCfgParams)
```

This way we will simplify the `generateUpstream` function considerably, so its signature can look like this:
```go
func generateUpstream(upstreamName string, upstream conf_v1alpha1.Upstream, endpoints []string, isExternalNameSvc bool, cfgParams *ConfigParams) version2.Upstream {
```

Note, we only have 5 parameters, as before this PR.

I believe in this case both the code and the test will be simplified.